torch/csrc/autograd/python_variable.cpp

```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
   END_HANDLE_TH_ERRORS
 }
 
+static py::list symint_array_to_list(SymIntArrayRef arr) {
+  py::list result(arr.size());
+  for (const auto idx : c10::irange(arr.size())) {
+    result[idx] = py::cast(arr[idx]);
+  }
+  return result;
+}
+
+static PyObject* DTensor_compute_global_tensor_info_impl(
+    const Tensor& tensor,
+    py::handle mesh,
+    const py::sequence& placements) {
+  Py_ssize_t idx = 0;
+  c10::SymDimVector tensor_shape(
+      tensor.sym_sizes().begin(), tensor.sym_sizes().end());
+  c10::SymDimVector tensor_strides(
+      tensor.sym_strides().begin(), tensor.sym_strides().end());
+  // NOTE: if this is a py::handle then this code stops working;
+  // apparently we can't rely on the bound method to stick around.
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify DeviceMesh somehow; profiling seems
+    // to say that nearly all our remaining time spent is spent
+    // calling back into Python.
+    const auto& cpp_placement = placement.cast<const distributed::Placement&>();
+    if (const auto* cpp_shard =
+            dynamic_cast<const distributed::Shard*>(&cpp_placement)) {
```

Comment:
I love the smell of RTTI for breakfast (do we really not have a type tag on these yet?)

Reply:
Hmm. It's been long enough that I'm not 100% sure what I was thinking, but.I now note that if this was just an  `is_shard()` call and someone has written a Python class that is a subclass of Placement but not the C++ shard class and yet is_shard returns true for it, sadness would ensue